src/Helper/BasePath.php

```diff
@@ -4,48 +4,61 @@
 
 namespace Laminas\View\Helper;
 
-use Laminas\View\Exception;
+use Laminas\View\Exception\RuntimeException;
 
 use function ltrim;
 use function rtrim;
+use function sprintf;
 
 /**
  * Helper for retrieving the base path.
+ *
+ * @final
  */
 class BasePath extends AbstractHelper
 {
-    /**
-     * Base path
-     *
-     * @var string
-     */
+    use DeprecatedAbstractHelperHierarchyTrait;
+
+    /** @var string|null */
     protected $basePath;
 
+    public function __construct(?string $basePath = null)
+    {
+        $this->basePath = $basePath;
+    }
+
     /**
      * Returns site's base path, or file with base path prepended.
      *
      * $file is appended to the base path for simplicity.
      *
      * @param  string|null $file
-     * @throws Exception\RuntimeException
+     * @throws RuntimeException
      * @return string
      */
     public function __invoke($file = null)
     {
-        if (null === $this->basePath) {
-            throw new Exception\RuntimeException('No base path provided');
+        if ($this->basePath === null) {
+            throw new RuntimeException('No base path provided');
         }
 
-        if (null !== $file) {
-            $file = '/' . ltrim($file, '/');
+        if (! empty($file)) {
+            return sprintf(
+                '%s/%s',
+                rtrim($this->basePath, '/'),
+                ltrim($file, '/')
+            );
         }
 
-        return $this->basePath . $file;
+        return rtrim($this->basePath, '/');
     }
 
     /**
      * Set the base path.
      *
+     * @deprecated since 2.21.0, this method will be removed in version 3.0.0 of this component.
```

Comment:
One design problem here: how does one run a long-running MVC application with this basePath fixed? 🤔

Reply:
I did consider the long running application scenario, because it's a target for all the work I'm doing. I personally think that if you're in a position where the removal of runtime changes to the base path affect you, then you've probably already written a custom helper to do this job. There's no existing mechanism to change the base path on a per-request basis. In MVC and Mezzio, it's the kind of thing that's done once and the presence of the method at all smells like the lack of DI from back in the ZF1 days.

Happy to revert, but I personally think that changing it during runtime is a routing concern, whereas the purpose of this helper is, as I understand it, to provide a stable prefix to _all_ routes or files. ??

Reply:
> I personally think that if you're in a position where the removal of runtime changes to the base path affect you, then you've probably already written a custom helper to do this job.

The goal cannot be for everyone to write their own helper when long-running applications are used.
So the question is if we can solve the problem here at all.

Reply:
I believe that the presence and purpose of the `setWhatever()` method here is about favouring setter injection in older versions of this component. My argument is that _if_ you are using View in a long running app, you are already in a whole world of hurt and you would have to already be dealing with all the state problems it brings. Moving it to the constructor and prohibiting runtime modification at least nails up the intentions of the original design (As I understand it) so that we can in future work out how view can be used successfully in long running apps. Maybe a clear definition of what base path is for would be helpful. Base path for static files? For everything? In its current state, I can't see how it would be a useful tool for multiple, different base paths, bearing in mind you get the same instance every time you fetch it from the container (By default)

Reply:
In this case, it's really not for setter injection, but because of the stateful nature of this helper.

Reply:
I investigated this in MVC and it looks like the setting is indeed only used at construction time in a factory [link](https://github.com/laminas/laminas-mvc/blob/ae0dbea9aed39dd6cb85f998b5ca4df320df90fd/src/Service/ViewHelperManagerFactory.php#L110)… but I've reverted the deprecation anyway as removing it doesn't really provide much benefit.